### DISCUSSION TITLE
How do I import freechips & rocket-chip?

### DISCUSSION THREAD

**Discussion by pepper-lees** [chipsalliance/rocket-chip#3394] [Q&A]
Hi,

I'm starting out with the [chisel template](https://github.com/freechipsproject/chisel-template) and I'm trying to make a RoCC accelerator. But when I try to import freechips.rocketchip (since I need to import LazyRoCC, LazyRoCCModuleImp, etc.):

```scala
import freechips.rocketchip.tile._
import freechips.rocketchip._
```

I get an error `not found: object freechips`. I tried searching through documentation on how to add freechips as a dependency to the project, but couldn't find anything. I tried copying the freechips-related configs from build.sbt in chipyard, too, but there's a ton of variables in there which I don't think are relevant to my project. What's the right way to import freechips?

**Comment by ZenithalHourlyRate** [comment]
This line might be relevant

https://github.com/ucb-bar/chipyard/blob/ca9e132bc61950f4d5e06b191316097a1642d641/build.sbt#L121

freechips is a package name, and things (e.g. directory name) in build.sbt can be irrelevant to the package name; When you added rocket-chip as your dependency, freechips package will be there.

**Reply by pepper-lees** [reply]
That's one of the blocks of code I tried copying. But note this:

`.dependsOn(hardfloat, rocketMacros, cde)`
so I also needed to copy over hardfloat and rocketMacros, which also depend on other variables, etc.

> When you added rocket-chip as your dependency, freechips package will be there.

Is there some command I can run to add rocket-chip as a dependency, instead of fumbling around with build.sbt?

**Reply by ZenithalHourlyRate** [reply]
https://github.com/ucb-bar/gemmini/blob/be2e9f26181658895ebc7ca7f7d6be6210f5cdef/build.sbt#L11

This might be a simpler solution

**Reply by pepper-lees** [reply]
@ZenithalHourlyRate Thanks, I think that's a lot closer!

After I added that line, it looks like it attempted to download from https://repo1.maven.org/maven2/edu/berkeley/cs/rocketchip_2.13/maven-metadata.xml
but after browsing that repo site, it looks like there's only https://repo1.maven.org/maven2/edu/berkeley/cs/rocketchip_2.12/ (2.12, not 2.13). Is there some way to auto-detect what version is available on the repo? (I don't understand how any of this works)

It's strange to me anyways, because I only see `"1.2.+"` in the version, so I would have guessed it should have detected the correct URL. Are there any references for this stuff works?

I tried to change "1.2.+" to "2.12" but it then attempted to build the url like this:
https://repo1.maven.org/maven2/edu/berkeley/cs/rocketchip_2.13/2.12/rocketchip_2.13-2.12.pom

**Reply by ZenithalHourlyRate** [reply]
> @ZenithalHourlyRate Thanks, I think that's a lot closer!
> 
> After I added that line, it looks like it attempted to download from https://repo1.maven.org/maven2/edu/berkeley/cs/rocketchip_2.13/maven-metadata.xml but after browsing that repo site, it looks like there's only https://repo1.maven.org/maven2/edu/berkeley/cs/rocketchip_2.12/ (2.12, not 2.13). Is there some way to auto-detect what version is available on the repo? (I don't understand how any of this works)
> 
> It's strange to me anyways, because I only see `"1.2.+"` in the version, so I would have guessed it should have detected the correct URL. Are there any references for this stuff works?

2.12 and 2.13 are scala version. Seems like you are using scala 2.13.

1.2.+ is rocketchip version. But sorry from my side, I forgot that after bumping to scala 2.13, rocketchip is never released to maven/sonatype, even though rocketchip has tagged 1.6 in github

You have to add rocketchip as submodule and build from source now, or you can choose to develop your accelerator in a forked rocketchip repo or in chipyard.
